File: src/test/java/tfol/typecheck/TermStructureTypeCheckTest.java
import static org.junit.Assert.assertEquals;
import org.junit.Test;
import org.junit.Ignore;

import fortress.tfol.*;
import java.util.List;

public class TermStructureTypeCheckTest {
    
    Sort A = Sort.mkSortConst("A");
    Sort B = Sort.mkSortConst("B");
    
    Var x = Term.mkVar("x");
    Var y = Term.mkVar("y");
    Var z = Term.mkVar("z");
    Var p = Term.mkVar("p");
    Var q = Term.mkVar("q");
    
    FuncDecl P = FuncDecl.mkFuncDecl("P", A, Sort.Bool());
    FuncDecl Q = FuncDecl.mkFuncDecl("Q", B, Sort.Bool());
    FuncDecl f = FuncDecl.mkFuncDecl("f", A, B);
    FuncDecl g = FuncDecl.mkFuncDecl("g", B, A);
    FuncDecl h = FuncDecl.mkFuncDecl("h", Sort.Bool(), Sort.Bool());
    FuncDecl R = FuncDecl.mkFuncDecl("R", A, A, Sort.Bool());
    
    // Term structure
    
    @Test(expected = fortress.data.TypeCheckException.BadStructure.class)
    public void forallInsideApp() {
        // Forall is not allowed inside an application
        Signature sig = Signature.empty()
            .withSorts(A)
            .withFunctionDeclarations(h);
        Term t = Term.mkApp("h", Term.mkForall(x.of(A), Term.mkTop()));
        t.typeCheck(sig);
    }
    
    @Test(expected = fortress.data.TypeCheckException.BadStructure.class)
    public void existsInsideApp() {
        // Exists is not allowed inside an application
        Signature sig = Signature.empty()
            .withSorts(A)
            .withFunctionDeclarations(h);
        Term t = Term.mkApp("h", Term.mkExists(x.of(A), Term.mkTop()));
        t.typeCheck(sig);
    }
    
    @Test(expected = fortress.data.TypeCheckException.BadStructure.class)
    public void connectiveInsideApp() {
        // Logical connectives are not allowed inside an application
        Signature sig = Signature.empty()
            .withFunctionDeclaration(h);
        Term t = Term.mkApp("h", Term.mkAnd(Term.mkTop(), Term.mkTop()));
        t.typeCheck(sig);
            
    }
    
    @Test(expected = fortress.data.TypeCheckException.BadStructure.class)
    public void negationInsideApp() {
        // Negation is not allowed inside an application
        Signature sig = Signature.empty()
            .withFunctionDeclaration(h);
        Term t = Term.mkApp("h", Term.mkNot(Term.mkTop()));
        t.typeCheck(sig);
    }
    
    @Test(expected = fortress.data.TypeCheckException.BadStructure.class)
    public void equalsInsideApp() {
        // = is not allowed inside an application
        Signature sig = Signature.empty()
            .withSort(A)
            .withConstant(x.of(A))
            .withFunctionDeclaration(h);
        Term t = Term.mkApp("h", Term.mkEq(x, x));
        t.typeCheck(sig);
    }
    
    @Test(expected = fortress.data.TypeCheckException.BadStructure.class)
    public void distinctInsideApp() {
        // distinct is not allowed inside an application
        Signature sig = Signature.empty()
            .withSort(A)
            .withConstant(x.of(A))
            .withFunctionDeclaration(h);
        Term t = Term.mkApp("h", Term.mkDistinct(x, x));
        t.typeCheck(sig);
    }
}

File: src/test/scala/msfol/typecheck/TermStructureTypeCheckTest.scala
import org.scalatest._

import fortress.msfol._
import fortress.transformers._
import fortress.operations.TermOps._

class TermStructureTypeCheckTest extends UnitSuite {
    
    val A = Sort.mkSortConst("A")
    val B = Sort.mkSortConst("B")
    
    val x = Var("x")
    val y = Var("y")
    val z = Var("z")
    val p = Var("p")
    val q = Var("q")
    
    val P = FuncDecl.mkFuncDecl("P", A, Sort.Bool)
    val Q = FuncDecl.mkFuncDecl("Q", B, Sort.Bool)
    val f = FuncDecl.mkFuncDecl("f", A, B)
    val g = FuncDecl.mkFuncDecl("g", B, A)
    val h = FuncDecl.mkFuncDecl("h", Sort.Bool, Sort.Bool)
    val R = FuncDecl.mkFuncDecl("R", A, A, Sort.Bool)
    
    // val structure
    
    test("forall inside app") {
        // Forall is not allowed inside an application
        val sig = Signature.empty
            .withSorts(A)
            .withFunctionDeclarations(h)
        val t = App("h", Forall(x.of(A), Top))
        an [fortress.data.TypeCheckException.BadStructure] should be thrownBy {
            t.typeCheck(sig)
        }
    }
    
    test("exists inside app") {
        // Exists is not allowed inside an application
        val sig = Signature.empty
            .withSorts(A)
            .withFunctionDeclarations(h)
        val t = App("h", Exists(x.of(A), Top))
        an [fortress.data.TypeCheckException.BadStructure] should be thrownBy {
            t.typeCheck(sig)
        }
    }
    
    test("connective inside app") {
        // Logical connectives are not allowed inside an application
        val sig = Signature.empty
            .withFunctionDeclaration(h)
        val t = App("h", And(Top, Top))
        an [fortress.data.TypeCheckException.BadStructure] should be thrownBy {
            t.typeCheck(sig)
        }
    }
    
    test("negation inside app") {
        // Negation is not allowed inside an application
        val sig = Signature.empty
            .withFunctionDeclaration(h)
        val t = App("h", Not(Top))
        an [fortress.data.TypeCheckException.BadStructure] should be thrownBy {
            t.typeCheck(sig)
        }
    }
    
    test("equals inside app") {
        // = is not allowed inside an application
        val sig = Signature.empty
            .withSort(A)
            .withConstant(x.of(A))
            .withFunctionDeclaration(h)
        val t = App("h", Eq(x, x))
        an [fortress.data.TypeCheckException.BadStructure] should be thrownBy {
            t.typeCheck(sig)
        }
    }
    
    test("distinct inside app") {
        // distinct is not allowed inside an application
        val sig = Signature.empty
            .withSort(A)
            .withConstant(x.of(A))
            .withFunctionDeclaration(h)
        val t = App("h", Distinct(x, x))
        an [fortress.data.TypeCheckException.BadStructure] should be thrownBy {
            t.typeCheck(sig)
        }
    }
    
    test("quantifier inside ite condition") {
        val sig = Signature.empty
            .withSort(A)
            .withConstant(x of A)
        val t = IfThenElse(Exists(y of A, y === y), x, x)
        an [fortress.data.TypeCheckException.BadStructure] should be thrownBy {
            t.typeCheck(sig)
        }
    }
}
